holochain/holochain-rust
# Tracing pt. 1

## PR summary

Note to self: see [tracing-1-michael](https://github.com/holochain/holochain-rust/tree/tracing-1-michael) for local crate paths for local dev on this

Adding basic tracing to core, using newly added updates to holochain-tracing.

- Sprinkles `#[autotrace]` macro in many places to build up full thread-local stack traces
- Liberally uses SpanWrap and EncodedSpanWrap to send Spans with data across thread and process boundaries, respectively.
    - action channel now uses SpanWrap
    - WireMessage lib3h communication uses EncodedSpanWrap
- Adds a Tracer and a long-running Reporting thread to both `holochain` and `cli` crates, which reports spans to a locally running Jaeger instance.
    - The conductor reporting thread dynamically creates a Reporter for each instance, and stores them all in a big RwLock<HashMap<_,_>>, similar to the signal receivers

## testing/benchmarking notes

### How to run tests locally:

Before running a test, run a Jaeger server (put this in a script)

```
docker run -d --rm --name jaeger \
  -p 5775:5775/udp \
  -p 6831:6831/udp \
  -p 6832:6832/udp \
  -p 5778:5778 \
  -p 16686:16686 \
  -p 14267:14267 \
  -p 14268:14268 \
  -p 9411:9411 \
  jaegertracing/all-in-one:1.16
```

- Make sure the test you want includes a "tracing" config in tryorama. See [app_spec/test/config.js](https://github.com/holochain/holochain-rust/pull/1925/files#diff-cf440d1063f2ba6dcb7fdb98f6157756R69) for an example
- Use the `--tracing-name` flag for sim2h_server, e.g. `hc-sim2h-server --tracing-name sim2h`
- Run the test, and go to http://localhost:16686 to see the Jaeger UI and search for traces

### Final exam / TryCP integration notes:

Run all-in-one on final exam:

```
docker run -d --rm --name jaeger \
  -p 5775:5775/udp \
  -p 6831:6831/udp \
  -p 6832:6832/udp \
  -p 5778:5778 \
  -p 16686:16686 \
  -p 14267:14267 \
  -p 14268:14268 \
  -p 9411:9411 \
  jaegertracing/all-in-one:1.16
```

Run jaeger-agent on each trycp machine

```
HOST=final-exam.holo.host

docker run \
  --rm \
  -p5775:5775/udp \
  -p6831:6831/udp \
  -p6832:6832/udp \
  -p5778:5778/tcp \
  jaegertracing/jaeger-agent:1.16 \
  --reporter.tchannel.host-port "$HOST:14267"
```


## followups

* Configurable sampling strategy
* More fine-tuned traces -- current ones are noisy and incomplete, in particular the discontinuity in queued_pending_validations, which is documented with TODOs in the code
* More use of tags
* Send traces over the network!

## changelog

- [ ] if this is a code change that effects some consumer (e.g. zome developers) of holochain core,  then it has been added to [our between-release changelog](https://github.com/holochain/holochain-rust/blob/develop/CHANGELOG-UNRELEASED.md) with the format 

```markdown
- summary of change [PR#1234](https://github.com/holochain/holochain-rust/pull/1234)
```

## documentation

- [x] this code has been documented according to our [docs checklist](https://hackmd.io/@freesig/Hk9AmKJNS)
